Amazon.Ion.ObjectMapper/IonSerializer.cs

```diff
@@ -286,84 +273,65 @@ public object Deserialize(IIonReader reader, Type type)
 
         public object Deserialize(IIonReader reader, Type type, IonType ionType)
         {
-            if (ionType == IonType.None || ionType == IonType.Null)
-            {
-                return new IonNullSerializer().Deserialize(reader);
-            }
-
-            if (ionType == IonType.Bool)
-            {
-                return new IonBooleanSerializer().Deserialize(reader);
-            }
-
-            if (ionType == IonType.Int)
-            {
-                if (reader.GetTypeAnnotations().Any(s => s.Equals(IonLongSerializer.ANNOTATION)))
-                {
-                    return new IonLongSerializer().Deserialize(reader);
-                }
-                return new IonIntSerializer().Deserialize(reader);
-            }
-
-            if (ionType == IonType.Float)
-            {
-                if (reader.GetTypeAnnotations().Any(s => s.Equals(IonFloatSerializer.ANNOTATION)))
-                {
-                    return new IonFloatSerializer().Deserialize(reader);
-                }
-                return new IonDoubleSerializer().Deserialize(reader);
-            }
-
-            if (ionType == IonType.Decimal)
+            try
             {
-                if (reader.GetTypeAnnotations().Any(s => s.Equals(IonDecimalSerializer.ANNOTATION)))
+                if (++currentDepth >= options.MaxDepth)
                 {
-                    return new IonDecimalSerializer().Deserialize(reader);
+                    throw new NotSupportedException($"Cannot deserialize further as max object tree depth {options.MaxDepth} is reached");

```

Comment:
Again this should not throw errors when it is exceeded

Reply:
Addressed.